src/bluetooth/BreachManager.js

```diff
@@ -0,0 +1,189 @@
+    return [!!configToCreateBreachFrom, configToCreateBreachFrom];
+  };
+
+  addLogToBreach = (breach, log) => {
+    const { id: temperatureBreachId } = breach;
+    return { ...log, temperatureBreachId };
+  };
+
+  willContinueBreach = (breach, log) => {
+    if (!breach) return false;
+    const { thresholdMaxTemperature, thresholdMinTemperature } = breach;
+    const { temperature } = log;
+
+    return temperature >= thresholdMinTemperature && temperature <= thresholdMaxTemperature;
+  };
+
+  willCloseBreach = (breach, log) => {
+    if (!(breach && log)) return false;
+    const { thresholdMaxTemperature, thresholdMinTemperature } = breach;
+    const { temperature } = log;
+    return !(temperature >= thresholdMinTemperature && temperature <= thresholdMaxTemperature);
+  };
+
+  couldBeInBreach = (log, configs) =>
+    configs.some(
+      config =>
+        log.temperature >= config.minimumTemperature && log.temperature <= config.maximumTemperature
+    );
+
+  createBreaches = (sensor, logs, configs, mostRecentBreach) => {
+    let potentialBreach = [];
+    let currentBreach = mostRecentBreach;
+
+    const breaches = mostRecentBreach && !mostRecentBreach.endTimestamp ? [mostRecentBreach] : [];
+    const temperatureLogs = [];
+
+    logs.forEach(log => {
+      const couldBeInBreach = this.couldBeInBreach(log, configs);
+      const willCloseBreach = this.willCloseBreach(currentBreach, log);
+      const willContinueBreach = this.willContinueBreach(currentBreach, log);
+
+      if (willCloseBreach) {
+        this.closeBreach(currentBreach, log.timestamp);
+        currentBreach = null;
+        potentialBreach = [];
+      }
+
+      if (couldBeInBreach) potentialBreach.push(log);
+      else potentialBreach = [];
+
+      if (willContinueBreach) {
+        const updatedLog = this.addLogToBreach(currentBreach, log);
+        temperatureLogs.push(updatedLog);
+      }
+
+      if (!willContinueBreach) {
+        const [willCreateBreach, config] = this.willCreateBreachFromConfigs(
+          configs,
+          potentialBreach
+        );
+        if (willCreateBreach) {
+          const newBreach = this.createBreach(
+            this.utils.createUuid(),
+            sensor,
+            config,
+            potentialBreach[0].timestamp
+          );
+          currentBreach = newBreach;
+          breaches.push(newBreach);
+          const updatedLogs = potentialBreach.map(l => this.addLogToBreach(newBreach, l));
+          updatedLogs.forEach(ul => temperatureLogs.push(ul));
+        }
+      }
+    });
+
+    return [breaches, temperatureLogs];
+  };
+
+  updateBreaches = async (breaches, temperatureLogs) => {
+    // eslint-disable-next-line no-param-reassign
+
+    const updatedBreaches = this.db.upsertBreaches(breaches);
+
+    const mapped = temperatureLogs.map(({ id, temperature, timestamp }) => ({
+      id,
+      temperature,
+      timestamp,
+    }));
+
+    // TODO: Copied from CCE app - not sure if needed here
```

Comment:
I think remove it - I should fix whatever the stupid thing I did in CCE app heh

Reply:
Cool, have removed - I guess we will find out when we call it if Realm does weird stuff with it? c553c4ad